### ISSUE TITLE
Poor handling of broken IPv6

### ISSUE THREAD

**Issue by mwoehlke-kitware** [python/cpython#105249]
# Bug report

On a system which produces IPv6 addresses for hostname lookups, but which is unable to communicate via IPv6, the following code:
```python
>>> import urllib.request
>>> urllib.request.urlopen('http://example.com')
```
...goes out to lunch for 130 seconds. (Substitute http://example.com for any valid HTTP URI whose host resolves to an IPv6 address.)

**Other programs** (e.g. `curl`, web browsers) don't have this problem, although it's unclear why. At a guess, other programs are obtaining both the IPv4 and IPv6 addresses and trying to connect to both simultaneously on a first-wins basis. (Network monitoring shows that e.g. curl sends an IPv4 SYN almost immediately after the IPv6 SYN. Python sends only the IPv6 SYN, followed by several TCP retransmissions, and does not send the IPv4 SYN until 130 seconds after the first IPv6 SYN.)

While this can be mitigated with timeouts (or by fixing the machine's network setup), it means that networking *in Python, specifically,* is essentially unusable.

If all applications were affected, it would be reasonable to chalk this up to bad network setup, but since other applications seem able to work around the issue with no user-visible effects, it would be preferable if Python could do likewise.

This also affects `requests`; there is some overlap at `http/client.py`, line 1329, `_send_request`, but the call stacks diverge both above and below that. Hopefully a fix is possible somewhere in the bowels of Python such that `requests` and other third-party users can also Just Work.

# Your environment

- CPython versions tested on: python3-3.11.3-2.fc37.x86_64
- Operating system and architecture: Fedora 37 x86_64

**Comment by terryjreedy** [comment]
`requests` is not part of the CPython stdlib and hence not the concern of this tracker.  Please give a reproducing example, similarly short if possible, using urllib.request and http.client.  Do they both use the same lower level module, so that the issue can be traced to a single bug?

**Comment by mwoehlke-kitware** [comment]
@terryjreedy, sorry. I've made the (fairly trivial) substitution. Whether there's a single point-of-"failure" is actually a good question and one of the reasons I mentioned both; they seem to be using `http.client` in subtly different ways.

I'm also not sure this is a "bug" in the sense that the code is doing something "definitely wrong". I suspect it's more a "bug" in the sense of a missing feature whose absence severely penalizes Python as compared to other software. (Hmm, that being the case, feel free to re-label if you think it's appropriate.)

**Comment by ronaldoussoren** [comment]
This is related to #88810, which would add the "Happy Eyeballs" algorithm to ``socket.create_connection`` (which is in the end the function used to create the TCP connection to the webserver).  Currently that function tries the various connection options serially which can result in the behaviour mentioned by the OP when IPv6 support is present but doesn't work properly.

**Comment by mwoehlke-kitware** [comment]
Note: pedantically, this appears to be an exact duplicate of #95838, which was also closed as a duplicate of #88810.
